Enhanced NMEA2000 String Handling with Unicode Support

davidhoy opened the pull request:
# Enhanced NMEA2000 String Handling with Unicode Support

## Overview

This pull request introduces comprehensive improvements to the NMEA2000 library, focusing on robust Unicode string handling and enhanced AIS message support. The changes ensure compliance with NMEA 2000 standards while providing backward compatibility and improved reliability.

## Key Features

### 🌐 Unicode and UTF-8 String Support

**Problem Solved**: The original library only supported ASCII strings, limiting internationalization capabilities and compliance with modern NMEA 2000 implementations that require Unicode support.

**Implementation**:

- **New UTF-8/UTF-16 Conversion Functions**: Added `Utf8ToUnicode()` and `UnicodeToUtf8()` functions for seamless conversion between UTF-8 and UTF-16LE encodings
- **Enhanced `AddVarStr()` Method**: Now supports both ASCII and Unicode strings with automatic format detection
- **SOH Header Convention**: Implements UTF-8 string identification using Start of Header (SOH, 0x01) prefix

### 📝 SOH Header Rationale

The SOH (Start of Header, ASCII 0x01) character is used as a prefix to indicate UTF-8/Unicode source strings for several technical reasons:

1. **Unambiguous Detection**: SOH is a non-printable control character that rarely appears in normal ASCII text, making it an excellent sentinel value
2. **NMEA 2000 Compatibility**: Preserves the existing binary format while extending functionality
3. **Backward Compatibility**: ASCII strings without SOH prefix continue to work as before
4. **Standard Practice**: SOH has historically been used as a delimiter/header in various communication protocols
5. **Minimal Overhead**: Single byte overhead for Unicode detection
6. **Clear Semantics**: When a string starts with SOH, the remainder is treated as UTF-8 and converted to UTF-16LE for NMEA 2000 Unicode transmission

### 🎯 Enhanced String Management

**New `tN2kStr` Class** (`N2kStr.h`):

- Unified string handling supporting both ASCII and Unicode formats
- Fixed-size buffer management with proper bounds checking
- Conversion methods between different string encodings
- Thread-safe Unicode string handling

**Improved `AddVarStr()` Method**:

- Added `maxAscii` and `maxUnicode` parameters for NMEA standards compliance
- Automatic length limiting based on NMEA 2000 field specifications
- Enhanced error handling and validation

### 📡 AIS Message Enhancement

**Sequence ID Support**: Added sequence ID parameter to `SetN2kPGN129802()` function for proper AIS message handling, enabling:

- Multi-part AIS message sequencing
- Improved AIS message reliability
- Better compliance with AIS standards

## Technical Details

### String Format Detection

```cpp
if (str[0] == 0x01) {
    // UTF-8/Unicode string detected - convert to UTF-16LE
    const char *utf8 = str + 1; // skip SOH
    // ... conversion logic
} else {
    // Standard ASCII string handling
}
```

### UTF-16LE Storage Format

- NMEA 2000 Unicode strings are stored in UTF-16LE (Little Endian) format
- Automatic conversion from UTF-8 input to UTF-16LE for transmission
- Proper byte ordering for cross-platform compatibility

### Length Limitations

- ASCII strings: Limited to 250 characters (configurable via `maxAscii`)
- Unicode strings: Limited to 125 characters (configurable via `maxUnicode`)
- Compliant with NMEA 2000 field length specifications

## Backward Compatibility

✅ **Full Backward Compatibility**: All existing ASCII string functionality remains unchanged  
✅ **API Compatibility**: Existing method signatures preserved with optional parameters  
✅ **Binary Compatibility**: No changes to existing message formats

## Testing and Validation

The changes have been tested with:

- ASCII string handling (existing functionality)
- UTF-8 to UTF-16LE conversion and vice versa
- NMEA 2000 message encoding/decoding
- Multi-language string support

## Standards Compliance

- **NMEA 2000**: Compliant with NMEA 2000 string field specifications
- **UTF-8/UTF-16**: Proper implementation of Unicode standards
- **AIS**: Enhanced AIS message sequence handling

## Files Modified

- `src/N2kMsg.cpp` - Core string handling and conversion logic
- `src/N2kMsg.h` - Method signatures and UTF conversion declarations  
- `src/N2kStr.h` - New unified string class (added)
- `src/NMEA2000.cpp` - Simplified PGN126998 implementation using new AddVarStr
- `src/N2kMessages.h/.cpp` - AIS sequence ID support

## Summary

This enhancement significantly improves the library's internationalization capabilities while maintaining rock-solid backward compatibility. The SOH header approach provides a clean, efficient method for distinguishing UTF-8 source strings while preserving the existing NMEA 2000 binary protocol structure.

davidhoy committed:
Fix some compiler warnings.

Commit:
Add support for UTF-8 and UTF-16 string conversions in NMEA2000 handling

- Introduced Utf8ToUnicode and UnicodeToUtf8 functions for string encoding conversions.
- Updated tN2kMsg class to handle variable-length ASCII and Unicode strings.
- Refactored SetN2kPGN126998 to utilize AddVarStr for improved string handling.
- Added new header file N2kStr.h for string management.

davidhoy committed:
Implement destructors for tN2kDeviceList and tNMEA2000 to manage dynamic memory cleanup

davidhoy committed:
Add Sequence ID parameter to SetN2kPGN129802 function for AIS messages

davidhoy committed:
Enhance AddVarStr method to include maxAscii and maxUnicode parameters to allow for limiting length to what is specified in the NMEA standards documentation.

ttlappalainen commented:
Why you need to use 0x01 as detection? Why just not use as I use bool RequireUnicode(const char *str) and then handle string?

```
// *****************************************************************************
static bool RequireUnicode(const char *str) {
  if ( str==0 ) return false;

  const unsigned char *bytes = (const unsigned char *)str;
  size_t num;
  bool ret=false;

    while ( *bytes != 0x00 ) {
      if ( (*bytes & 0x80) == 0x00 ) {
        // U+0000 to U+007F 
        num = 1;
      } else if ((*bytes & 0xE0) == 0xC0) {
        // U+0080 to U+07FF 
        num = 2;
      } else if ((*bytes & 0xF0) == 0xE0) {
        // U+0800 to U+FFFF 
        num = 3;
      } else if ((*bytes & 0xF8) == 0xF0) {
        // U+10000 to U+10FFFF 
        num = 4;
      } else if ( (*bytes & 0xFC) == 0xF8 ) {
        num = 5; 
      } else if ( (*bytes & 0xFE) == 0xFC ) {
        num = 6; 
      } else return false;

      bytes++;
      for ( int i = 1; i < num; ++i ) {
        if ( (*bytes & 0xC0) != 0x80) return false;
        bytes++;
      }

      if ( num>2 ) return false; // Can not handle
      if ( num==2 ) ret=true;
  }

  return ret;
}
```

It shold be enough to have void tN2kMsg::AddVarStr(const char *str, bool UsePgm, size_t maxAscii=0);  maxAscii=0 that all available can be used. maxUnicode is half of that. 250 is not good as default.

In PGN 1261998 setter you can not just set strings provided. They must be limited to 70.

Instead of , uint8_t SID=255 there could be , uint8_t SID=N2kUInt8NA

davidhoy commented:
Hi Timo,

The reason for using the SOH prefix for Unicode strings that have been re-encoded to UTF-8 is so that the embedded application does not need to explicitly support Unicode or UTF-8.  The string is simply treated as an ASCII string by the application.  Imagine this scenario - the end user changes the installation labels, using a tool such as Maretron's N2KAnalyzer, and uses Unicode but with only ASCII characters.  The app processes the PGN, and converts the string to UTF-8 with the SOH prefix.  Then later the 126998 PGN is requested, and the stack seems that the string has the SOH prefix, so converts to Unicode for the response PGN.  Without the SOH prefix it would send it back as ASCII, which is not correct. 

Regarding the maxAscii and maxUnicode parameters, I believe it give the application the most flexibility, in case of future changes.  Most callers will never even know.  Just using half is very inflexible.  As to the limits in PGN 126998, I'm looking at the NMEA spec right now, and it says 250/125 for all three fields - installation 1, installation 2, and manufacturer info.

Agreed on the SID=255, I'll change that to N2KUInt8NA...

-David

davidhoy committed:
Update SID default value in SetN2kPGN129802 and SetN2kAISSafetyRelatedBroadcastMsg functions to N2kUInt8NA for improved clarity.  As recommended by Timo L.

davidhoy committed:
Enhance SetN2kPGN126998 function to limit Installation Description and Manufacturer Information lengths to 70 ASCII or 35 Unicode characters.

ttlappalainen commented:
I do not see anything wrong that if string has been set as unicode, but it does nooed to be unicode, it just saves ascii and returns it as ascii later. Is there some definition for that?

In my own code I simply get e.g., InstallationDescription1 and convert it to UTF8. Then I save received string. When it will be requested, I'll return it as ascii, if there is no need to use unicode.

ttlappalainen commented:
If you have UTF8 string and you use on AddVarStr RequireUnicode for testing, it will return ascii or unicode, if required. I have not failed with that in tests. Could you point test which have requirement to return unicode even ascii is could be used, so I'll recheck.

ttlappalainen commented:
3.10.4 does not match.

Anyway I do not think that this is proper way. String is string and not meant to carry extra information how it should be handled in specially NMEA2000. With this path there could be special markers from different libraries. It would be best to handle as I mentioned strings as ascii or UTF-8 and use RequireUnicode.

About tests there has been other errors too and they have been accepted with explanation. Those tests has been fixed in newer test version. Also newer test has some  inconsistent. So I do not see any reason to carry string as unicode, if it can be carried as ascii.

davidhoy commented:
Hi Timo, thanks for continuing the discussion — let me clarify why I think the SOH prefix and the additional parameters are necessary.

### 1. Certification compliance  
The key driver here is **NMEA 2000 certification**. The test suite (e.g., *3.10.4*) explicitly sets Unicode strings and expects them to be returned as Unicode, even when they contain only ASCII characters. Without a marker, those strings get silently downgraded and returned as ASCII, which causes certification failures. This isn’t theoretical — I’ve reproduced it with N2KAnalyzer and the certification toolchain. So the library as-is cannot reliably pass certification in this case.

### 2. Limits of UTF-8 detection  
Your `RequireUnicode()` detection works well when the string contains multi-byte UTF-8 sequences. But when the string consists only of ASCII characters encoded as Unicode, there’s no way to tell whether it was intended as ASCII or Unicode. That distinction matters, because certification requires round-tripping the *intent* of the string, not just the code points. Without an explicit marker, the “Unicode-ness” is lost.

### 3. Why SOH works here  
SOH (`0x01`) is an ideal sentinel:  
- It never appears in printable NMEA 2000 strings, so there’s no collision risk.  
- It adds just one byte of overhead.  
- It leaves all existing ASCII behavior intact.  
- Applications that don’t care about Unicode still treat the string as plain ASCII.  

This isn’t arbitrary — it’s a simple, pragmatic way to preserve intent through the full encode/decode cycle, which pure detection cannot do.

### 4. Avoiding hidden failures  
The biggest risk is a **silent downgrade**: developers assume Unicode strings are being preserved, but in reality they come back as ASCII. That kind of hidden bug only shows up at certification time or in the field with third-party tools, which is far more painful than carrying a single byte of metadata.

### 5. Flexibility in `AddVarStr`  
The extra `maxAscii` and `maxUnicode` parameters give us flexibility for both the `250/125` and `70/35` limits that appear in the spec. Hard-coding “half of ASCII length” is brittle. Most callers never override the defaults, but for those that do, this keeps the library compliant without local patches.

### 6. Respectfully addressing concerns  
I agree in principle that *“a string is just a string”*, but NMEA 2000 already uses metadata everywhere (binary headers, PGNs, instance IDs). One sentinel byte is consistent with that model and solves a real certification gap.  

And while test versions have evolved, supporting correct Unicode round-tripping is harmless and future-proof. If it helps, I’m happy to make the SOH behavior optional (compile-time or runtime flag) so projects that don’t want it can disable it.

---

**In short:** the SOH sentinel isn’t about complicating the library — it’s about ensuring devices using this library pass certification and interoperate correctly. It solves a real ambiguity that pure detection can’t handle, with negligible overhead and full backward compatibility. Without it, developers will end up writing ad-hoc patches, which fragments the ecosystem.  

My goal is to strengthen this library so it remains the go-to reference for certified NMEA 2000 implementations.

ttlappalainen commented:
1. his is definetely wrong and should be opened as question to organization. There should not be any reason to require unicode, if ascii can be used.
2.  I prefer AddVarStr should transparently support ascii/UTF-8. For setting unicode there could be own function or parameter.
3. As I mentioned string should not carry information how to handle it. Classes are for that purpose.
4. Again classess.
5. I still do not see use for maxUnicode. For now it is always half of maxAscii and could be actually defined as maxChars.
6. Yes NMEA2000 uses metadata, but API does not need to.

davidhoy committed:
Merge remote-tracking branch 'refs/remotes/origin/master'

ttlappalainen commented:
Which test tool version you have? Which version of documents you have?

As I have mentioned I would prefer just ascii and UTF-8 support. UTF-8 is commonly used in html and json. If you can not point method for marking unicode string with starting byte 01 as standard, it is hack and not good idea.

Also if some test set string as unicode string and expects unicode back even it can be saved ascii, it has no point. Some other may set same string as ascii and devices will no expect to get it as unicode. There has been similar errors in test over time and they have been accepted with explanation.

ttlappalainen commented:
I do not understand your sentence "that distinction matters, because certification requires round-tripping the intent of the string". At what point of documentation thay say that, if string has been set with unicode event it can be translated to ascii, it should be returned as unicode?

ttlappalainen commented:
I use different code with UTF8 support, which automatically uses then Unicode string. I'll implement that code to published library.

davidhoy committed:
Add address claim and send frame handler callbacks to NMEA2000 class

ttlappalainen commented:
This is now outdated, since library seamlessly converts UTF8 strings to unicode if necessary. NMEA organization has confirmed that unicode is not mandatory and C.3.22.9 can be handle with verification.

For new callbacks I am thinking of method mentioned on https://github.com/ttlappalainen/NMEA2000/pull/404.
